Sources/SemanticIndex/UpdateIndexStoreTaskDescription.swift

```diff
@@ -475,29 +634,89 @@ package struct UpdateIndexStoreTaskDescription: IndexTaskDescription {
+        // BuildSettingsLogger.log(level: .error, settings: buildSettings, for: indexFile)
       }
     case .abnormal(let exception):
       if !Task.isCancelled {
-        logger.error("Updating index store for \(indexFile.forLogging) exited abnormally \(exception)")
-        BuildSettingsLogger.log(level: .error, settings: buildSettings, for: indexFile)
+        logger.error("Updating index store exited abnormally \(exception) for \(indexFiles)")
+        // BuildSettingsLogger.log(level: .error, settings: buildSettings, for: indexFile)
       }
     }
   }
+
+  /// Partition the given `FileIndexInfos` into batches so that a single `UpdateIndexStoreTaskDescription` should be
+  /// created for every one of these batches, taking advantage of multi-file indexing when it is supported.
+  static func batches(
+    toIndex fileIndexInfos: [FileIndexInfo],
+    buildServerManager: BuildServerManager
+  ) async -> [(target: BuildTargetIdentifier, language: Language, files: [FileIndexInfo])] {
+    struct TargetAndLanguage: Hashable, Comparable {
+      let target: BuildTargetIdentifier
+      let language: Language
+
+      static func < (lhs: TargetAndLanguage, rhs: TargetAndLanguage) -> Bool {
+        if lhs.target.uri.stringValue < rhs.target.uri.stringValue {
+          return true
+        } else if lhs.target.uri.stringValue > rhs.target.uri.stringValue {
+          return false
+        }
+        if lhs.language.rawValue < rhs.language.rawValue {
+          return true
+        } else if lhs.language.rawValue > rhs.language.rawValue {
+          return false
+        }
+        return false
+      }
+    }
+
+    var partitions: [(target: BuildTargetIdentifier, language: Language, files: [FileIndexInfo])] = []
+    var fileIndexInfosToBatch: [TargetAndLanguage: [FileIndexInfo]] = [:]
+    for fileIndexInfo in fileIndexInfos {
+      guard fileIndexInfo.language == .swift,
+        await buildServerManager.toolchain(for: fileIndexInfo.target, language: fileIndexInfo.language)?
+          .canIndexMultipleSwiftFilesInSingleInvocation ?? false
+      else {
+        // Only Swift supports indexing multiple files in a single compiler invocation, so don't batch files of other
+        // languages.
+        partitions.append((fileIndexInfo.target, fileIndexInfo.language, [fileIndexInfo]))
+        continue
+      }
+      // Even for Swift files, we can only index files in a single compiler invocation if they have the same build
+      // settings (modulo some normalization in `updateIndexStore(forFiles:)`). We can't know that all the files do
+      // indeed have the same compiler arguments but loading build settings during scheduling from the build server
+      // is not desirable because it slows down a bottleneck.
+      // Since Swift files within the same target should build a single module, it is reasonable to assume that they all
+      // share the same build settings.
```

Comment:
I'm not sure I'd even call this an assumption TBH - if it wasn't the case, things would be very broken.

Reply:
Same reply as https://github.com/swiftlang/sourcekit-lsp/pull/2293#discussion_r2365638192. I guess it’s just nice to see that SourceKit-LSP continues to function even if the build server gives us unreasonable responses.